man/audiomass.1
.TH AUDIOMASS 1
.SH NAME
.B audiomass
\- Wrapper for command line audio libraries.
.SH SYNOPSIS

.B audiomass
\fI[-h HELP] [-v VERSION] [-c COPYING] [-C CHECK]
[-f FILE] [-d DIRECTORY] [-b BATCH [..FILE1 ..FILE2 ..FILE3 ..]] [-o DIRNAME]\fR

.SH DESCRIPTION

Audiomass is a command line audio wrapper of the Flac, Lame, Vorbis-tools,
Monkey's Audio, Shntool and FFmpeg libraries.
It supports conversions of different audio formats at a time and the ability
to convert even groups of files in a directory, saving the output in a specific
folder.

For each input file, audiomass-cli writes to a filename based on the input
filename, but with the extension changed to as appropriate to encoding.

.SH OPTIONS
.TP
.B \-f, \-\-file
Single audio conversion. If the (\fB\-o\fR) option is not specified, write to
the same source directory.
.TP
.B \-d, \-\-dir
Converts a bunch of audio files with the same extension contained in a directory.
If the (\fB\-o\fR) option is not specified, write to the same source directory.
.TP
.B \-b, \-\-batch
Load a queue of audio files even with different formats located in different
locations. If option (\fB\-o\fR) is not specified, writes the output to same
input file. If more than one file is specified, they must either be separated
to one or more white-space.
.TP
.B \-o \-\-output
Save the output files into specified folder. Can be used in conjunction with
the (\fB\-f\fR), (\fB\-d\fR) and (\fB\-b\fR) options described above. If you
want to save the outputs to a different folder, you might consider using this
option.
.TP
.B \-C \-\-check
This option checks the required dependencies by audiomass.
Installed packages are marked with '..Ok', while non-installed packages are
marked with '..Not Installed'
.TP
.B \-h, \-\-help
Show a short help and exit.
.TP
.B \-v, \-\-version
Show version and date and exit.
.TP
.B \-c, \-\-copying
Show license and exit.

.SH EXAMPLES
.TP
Convert a single file and save it into same directory:
.TP
.B audiomass-cli -f '/MyDirName/SomeFile.wav''
.TP
---------------------
.TP
Convert a bunch of audio files inside a folder and save them in another folder:
.TP
.B audiomass-cli -d /MyDirName/Music  -o '/MyOtherDir/converted
.TP
---------------------
.TP
Convert a queue of audio files and save the output in a specified folder:
.TP
.B audiomass -b '..STREAM1' '..STREAM2' '..STREAM3' '...' -o /output/dir

.SH "SEE ALSO"
.BR flac (1),
.BR oggenc (1),
.BR oggdec (1),
.BR lame (1),
.BR shntool (1),
.BR wavpack (1)
.SH AUTHORS
Written by Gianluca Pernigotto, Alias: jeanslack <jeanlucperni@gmail.com> .
Homepage: https://github.com/jeanslack/audiomass
.SH COPYRIGHT
Copyright  ©  2015-2019  Free Software Foundation, Inc.  License GPLv3+: GNU
GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free  to  change  and  redistribute  it.
There is NO WARRANTY, to the extent permitted by law.


CHANGELOG
+--------------------------+
Aug. 31 2020 Version 0.8.1
+--------------------------+

* Removed constant definitions from modules. Used __init__.py to definitions.
* Fixed setup.py to use setuptools.
* Fixed PEP8 with pycodestyle.
* Fixed bug on process_dir.py (glob.glob) when converts files on directory.
* The Audiomass Python package has been renamed to src


+--------------------------+
Dec. 28 2017 Version 0.7.6
+--------------------------+

- Add aiff conversions support for MP3 and OGG formats.
- Add wav, aiff, mp3 and ogg support for APE format.
- Add 'get()' method in comparisions.py
- Resolved bug in process_batch.py. In 'for' loop raise a [RuntimeError:
  dictionary changed size during iteration] when turn with 'continue' statement.
- Add a limites formats list into comparisions.supported_formats() dictionary and
  removed some redundant code into all process files
- Some improved console output
- Removed the whichcraft retro-compatibility with python 2, it used
  *shutil.which* module instead of


+--------------------------+
Dec. 17 2017 Version 0.7.5
+--------------------------+

 - Add compatibility for python3
 - Change the setup.py for build audiomass deb package for python3
 - Change debian dependencies names for building tools, see: README.MD
 - Change the command names for building package, see: README.md
 - Add new option for check installed dependencies
 - Re-formated datastrings.py content
 - Re-created new manual in pdf format
 - Resolved shntool 'ask option' for overwriting existance file.
   For now all existance file are *overwritings* with shntool only.
 - When load file with no extension with -f option result:
   audiomass: Error! Not format supported: ""
   Resolve with 'os.path.basename'
 - Resolved bug into parser if missing arg after -b option
 - Some tweaks to the code for a future backward compatibility with python3

+--------------------------+
Dec. 15 2017 Version 0.7.4
+--------------------------+

 - Add new option for check installed dependencies
 - Re-formated datastrings.py content
 - Re-created new manual in pdf format
 - Resolved shntool 'ask option' for overwriting existance file.
   For now all existance file are overwritings with shntool only.
 - When load file with no extension with -f option result:
   audiomass: Error! Not format supported: ""
   Resolve with 'os.path.basename'
 - Resolved bug into parser if missing arg after -b option

+--------------------------+
Dec. 06 2017 Version 0.7.3
+--------------------------+

 - Add functionalities for resolving some issue in the user arguments;
   Normalize a pathnames by collapsing redundant separators and up-level
   references. So you can write '/home/john/MyMusic/' or ''/home/john/MyMusic''
   or /home/john/MyMusic and return even '/home/john/MyMusic'.
   This avoid many others errors at the end of the process
 - Re-created new manual pdf format
 - Resolved shntool 'ask option' for overwriting existance file.
   For now all existance file are *overwritings* with shntool only.

+--------------------------+
Dec. 06 2017 Version 0.7.2
+--------------------------+

 - Now every file to be converted can have either uppercase or
   lowercase names
 - Fixed the bugs of the previous version
 - The code has also been improved a little bit

+--------------------------+
Dec. 01 2017 Version 0.7.1
+--------------------------+

NOTE: this is a definitive release.

 - New batch process implementation for convert multiple audio stream.
   The limits of the previous version are now outdated and work.
 - I continued with the improvements in errors management
 - Add new batch process option for files conversion with
   different formats :-)
 - New process management for single file, files in dir and batch
   conversions.
 - Add new implementations for saving in different folder for all
   the available process: single file conversion, file group conversion
   on directory and batch conversions.
 - Now markeup cosmetic are more beautiful
 - better organization of the code

+--------------------------+
November 30 2017 Version 0.7.0
+--------------------------+

NOTE: this release is just a passing version to experiment with some
      variations. Although it is usable it may have some bugs
      and generally it does not work very well.

- I continued with the improvements in errors management
- Add new batch process option for files conversion with
  different formats :-)
  This has some limits on output formats choice with -b option.
- New process management for single file, files in dir and batch
  conversions.
- Add new implementations for saving in different folder for all
  the process
- some small cosmetic makeup

+--------------------------+
November 22 2017 Version 0.6.1
+--------------------------+

- some cosmetic touch to the exit code
- Improved errors handling, but there is still work

+--------------------------+
febbruary 24 2015 Version 0.6
+--------------------------+

Initial release


src/audiomass.py
# -*- coding: utf-8 -*-
#
#########################################################
# Name: audiomass.py
# Porpose:  Boot straps and arg parser for audiomass script
# Writer: Gianluca Pernigoto <jeanlucperni@gmail.com>
# Copyright: (c) 2015/2017 Gianluca Pernigoto <jeanlucperni@gmail.com>
# license: GPL3
# Version: (Ver.0.7) December 2017
# Rev: dec 9 2017, Aug 8 2019
#########################################################
import sys
import os
from src import (
    __author__,
    __mail__,
    __copyright__,
    __version__,
    __release__,
    __rls_name__,
    __prg_name__,
    __url__,
    __short_descript__,
    __long_descript__,
    __usage__,
    __try__,
    __license__,
    __short_license__,
    )
from src.datastrings import msg_str
from src.whichcraft import check_dependencies
from src.process_file import file_parser
from src.process_dir import dir_parser
from src.process_batch import batch_parser


def main():
    """
    Boot straps and arg parsing

    """
    msg = msg_str()
    title = ("\033[1m%s\033[0m - Wrapper for multiple audio "
             "conversion libraries." % (__rls_name__))

    if '-f' in sys.argv or '--file' in sys.argv:
        print(title)
        get_file()
    elif '-d' in sys.argv or '--dir' in sys.argv:
        print(title)
        get_dir()
    elif '-b' in sys.argv or '--batch' in sys.argv:
        print(title)
        get_batch()
    elif '-C' in sys.argv or '--check' in sys.argv:
        print(title)
        check_dependencies()
    elif '-v' in sys.argv or '--version' in sys.argv:
        print("%s - version %s - released %s" % (__rls_name__,
                                                 __version__,
                                                 __release__
                                                 ))
    elif '-h' in sys.argv or '--help' in sys.argv:
        print(__usage__)
    elif '-c' in sys.argv or '--copying' in sys.argv:
        print(__license__)
    elif len(sys.argv) == 1:
        print(title)
        print(__try__)
    else:
        raise SystemExit("%s Invalid options: '%s' \n%s" % (msg[1],
                                                            sys.argv[1:][0],
                                                            __try__
                                                            ))
    return 0


def get_file():
    """
    Get the filename and output dirname if output dirname in sys.argv

    """
    msg = msg_str()
    opts = [x for x in ['-o', '--output'] if x in sys.argv]
    try:
        if os.path.isfile(sys.argv[2]):  # must be file only here
            path_in = os.path.abspath(os.path.join(sys.argv[2]))  # filename
        else:
            sys.exit("%s %s" % (msg[2], sys.argv[2]))
    except IndexError:
        sys.exit("%s Missing argument after option '%s'\n%s" % (msg[1],
                                                                sys.argv[1],
                                                                __try__
                                                                ))
    if len(sys.argv) == 4 or len(sys.argv) >= 4:
        # if sys.argv[3] != '-o':
        if sys.argv[3] not in opts:
            raise SystemExit("%s Invalid option: '%s' \n%s" % (msg[1],
                                                               sys.argv[3],
                                                               __try__
                                                               ))
        if len(sys.argv) >= 5:
            if os.path.isdir(sys.argv[4]):
                path_out = os.path.abspath(os.path.join(sys.argv[4]))  # dirn
            else:
                raise SystemExit("%s %s" % (msg[3], sys.argv[4]))
        else:
            raise SystemExit("%s Missing output dir-name after option "
                             "'%s'\n%s" % (msg[1], opts[0], __try__))
    else:
        path_out = None
    # input_format is the extension format of path_in
    input_format = os.path.splitext(path_in)[1].replace(".", "")
    file_parser(input_format, path_in, path_out)


def get_dir():
    """
    Get input dirname and output dirname if output dirname in sys.argv

    """
    msg = msg_str()
    opts = [x for x in ['-o', '--output'] if x in sys.argv]
    try:
        if os.path.isdir(sys.argv[2]):
            path_in = os.path.abspath(os.path.join(sys.argv[2]))  # dirname
        else:
            sys.exit("%s %s" % (msg[3], sys.argv[2]))
    except IndexError:
        sys.exit("%s Missing argument\n%s" % (msg[1], __try__))

    if len(sys.argv) == 4 or len(sys.argv) >= 4:
        if sys.argv[3] not in opts:
            raise SystemExit("%s Invalid option: '%s' \n%s" % (msg[1],
                                                               sys.argv[3],
                                                               __try__
                                                               ))
        if len(sys.argv) >= 5:
            if os.path.isdir(sys.argv[4]):
                path_out = os.path.abspath(os.path.join(sys.argv[4]))  # dirn
            else:
                raise SystemExit("%s %s" % (msg[3], sys.argv[4]))
        else:
            raise SystemExit("%s Missing output dir-name after option "
                             "'%s'\n%s" % (msg[1], opts[0], __try__))
    else:
        path_out = None
    dir_parser(path_in, path_out)


def get_batch():
    """
    Groups a queued input file stream and puts it in the queue list.
    Call the specified function and send queue list and output dirname
    if output dirname in sys.argv.

    """
    msg = msg_str()
    queue = []
    opts = [x for x in ['-o', '--output'] if x in sys.argv]
    if opts:
        opt = sys.argv.index(opts[0])
        try:
            if sys.argv[opt + 1]:  # se trovo percorso dopo opzione -o --output
                path_O = sys.argv[opt + 1]

        except IndexError:
            sys.exit("%s Missing argument after option '%s'\n%s" % (msg[1],
                                                                    opts[0],
                                                                    __try__
                                                                    ))
        else:
            if os.path.isdir(path_O):  # se opzione e percorso output corretti
                arg = sys.argv[2:]  # incorpora solo gli input pathnames
                arg.remove(opts[0])  # infatti rimuovo '-o' , '--output'
                arg.remove(path_O)  # e rimuovo l'eventuale output pathname
                path_O = os.path.abspath(os.path.join(path_O))  # pathname set
            else:
                raise SystemExit("%s %s" % (msg[3], path_O))
    else:
        path_O = None
        arg = sys.argv[2:]  # dal 2° arg. lista

    for f in arg:
        if os.path.isfile(os.path.abspath(os.path.join(f))):  # must be file
            queue.append(f)
        else:
            raise SystemExit("%s '%s'" % (msg[2], f))
    if not queue:
        raise SystemExit("%s Missing argument after option '%s'\n%s" % (
                                                                msg[1],
                                                                sys.argv[1],
                                                                __try__
                                                                ))
    batch_parser(queue, path_O)


if __name__ == '__main__':
    status = main()
    sys.exit(status)
